### PR TITLE
[Feature] Delete empty files

### PR DESCRIPTION
**Pull request by marksto**
Hi Michiel!

Here's a new option implementation which covers #72. It's called `:delete-empty-files` and for now it will only work for those files that got some of their vars carved. If a namespace was empty, i.e. there were no vars to carve, it shall remain, even if the namespace is itself unused. This is intentional.

There are a few legit cases for having empty namespaces I can think of, e.g. to require other namespaces (a fairly common pattern). And these empty namespaces may also seem unused for whatever quirky reason.

Since it touches way more ground, we can add a feature for deleting files with empty unused namespaces later, if needed. At minimum, it will require us updating the reporting and, most importantly, arranging it's own ignorance support that is to play nicely with existing `.ignore` logic for vars. 

Moreover, this full-fledged empty namespaces carving feature will kinda widen the scope of the tool, which at the moment only removes unused vars. Still, information about unused namespaces is at our fingertips, since `clj-kondo` provides it as well.

```
{:keys [:namespace-definitions :namespace-usages :var-definitions :var-usages]} (:analysis result)
unused-nss (set/difference (set (map :filename namespace-definitions))
                           (set (map :filename namespace-usages)))
```

Anyway, please let me know what you think about the "namespaces carving".

### CODE REVIEW

File: src/carve/impl.clj [hunk 1]
```diff
@@ -173,19 +180,46 @@
...
+                                      (some-fn ns-declaration? comment-form?)
+                                      zloc)))]
+      (when empty?
+        (when-not silent (println "Deleting empty file" (.getCanonicalPath file)))
+        (io/delete-file file)))
+    (catch Exception e
+      (when-not silent
+        (binding [*out* *err*]
+          (println (str "Exception thrown when analyzing and/or deleting " file "."))
+          (println e))))) )
```
Comment (marksto) [inline src/carve/impl.clj:221]: Here's the main logic. Quite straightforward thanks to `rewrite-clj.zip`.

Comment (marksto) [inline src/carve/impl.clj:221]: Here's the main logic. Quite straightforward thanks to `rewrite-clj.zip`.

File: src/carve/impl.clj [hunk 2]
```diff
@@ -277,12 +313,14 @@
             results (reduce into results [unused-vars-data unused-var-refers])]
         (if (or (seq unused-vars-data)
                 (seq unused-var-refers))
-          (do (when-not (:report opts)
+          (do (when-not report
```
Comment (marksto) [inline src/carve/impl.clj:316]: This one was not enlisted within other `:keys [...]` for some reason.

**Comment by borkdude** [comment]
I'll take a look at this later this week

**Comment by borkdude** [comment]
> for now it will only work for those files that got some of their vars carved

> e.g. to require other namespaces 

So this gets a little bit more complex than I anticipated.
I'd say if the namespace requires other namespaces, it's not safe to assume it can be removed, unless carve/kondo can prove that those namespaces are itself unused. Also you have to ensure that the namespace of the file itself isn't used anywhere else. This can get quite tricky. As such, I'm not sure if I'm comfortable including (and thus supporting + maintaining) this in carve yet, unless you can convince me otherwise :)

**Comment by marksto** [comment]
> ... unless you can convince me otherwise :)

Sure, that is exactly what I am going to do. =)

See, the current implementation is completely comfortable to have in Carve, since it stops before all these red lines — we only delete a namespace file in case **all its contents are carved out**. It defines nothing (there is only an `ns` form in its file).

Such a namespace:
- [ ] won't be used in other namespaces, since it's empty and there's nothing to refer to in it
- [ ] won't have any any side-effects and/or affect any global state (by defining multi-methods, specs, etc.), since it's empty and there's no code to trigger this
- [ ] can be "used anywhere else" by some other means (e.g. via some quirky strings + requiring-resolves manipulations) though, but this is: a) a quite rare case, b) a good signal for a user to update their code, and c) it could still be ignored (*).

That being said, we can still discuss the last point (*). Should I also implement this empty file ignore list within the current PR? Or we can address this corner case later on, whenever there is a real need for this?

Cheers,
Mark

**Comment by borkdude** [comment]
I've looked at the code again, but it's still not obvious that the a file won't be removed if it still loads other namespaces, like so:

``` clojure
(ns foobar.baz
  (:require [foobar.specs] ;; this must be loaded
  ))

(defn this-is-unused-so-carve-away! [])
```

Can you point to the section in the code that is responsible for not removing that file?

**Comment by borkdude** [comment]
When I tested this, it did remove that file:

```
$ bb -m carve.main --paths /tmp/carve_test/foobar --delete-empty-files
Carving /tmp/carve_test/foobar/baz.clj

Found unused var:
------------------
(defn this-is-unused-so-carve-away! [])
------------------
Type Y to remove or i to add foobar.baz/this-is-unused-so-carve-away! to .carve/ignore
Y
Writing result to /private/tmp/carve_test/foobar/baz.clj
Deleting empty file /private/tmp/carve_test/foobar/baz.clj
```

**Comment by marksto** [comment]
Just for the record then.

This is why I proposed to also augment the existing `.carve/ignore` file format — to be able to enlist there namespaces that must not be deleted in no case. The same way as you already do it in a "less risky" vars carving scenario.

And I feel the aforementioned case — to have a namespace that does nothing but require other namespaces (since all its vars were unused and got carved) — a quite rare one. This may happen, of course, but, say, in less than 5% of all cases, even if you have a vast monorepo codebase with hundreds of namespaces.

In my specific case I will to use Carve with a bloated internal utility library which is reused across multiple other projects. I know that each project uses just a handful of its namespaces and functions, thus everything else can be safely removed — vars should get carved and empty files deleted. There are no specs and/or other stuff to worry about. We just need to clean up. This case is from the other 95%, as I see it.

That said, I'll keep my fork and live with it. Cheers!

**Comment by borkdude** [comment]
@marksto I will re-consider if more users are interested in this feature and it comes up more often.

**Pull request closed without merging**